source/collectBy.js

```diff
@@ -0,0 +1,37 @@
+import _curry2 from './internal/_curry2';
+import groupBy from './groupBy';
+import values from './values';
+
+/**
+ * Splits a list into sub-lists, based on the result of calling a String-returning function on each element,
+ * and grouping the results according to values returned.
+ *
+ * @func
+ * @memberOf R
+ * @category List
+ * @sig (a -> String) -> [a] -> {String: [a]}
+ * @param {Function} fn Function :: a -> String
+ * @param {Array} list The array to split into sub-arrays.
+ * @return {Array}
+ *    An array of arrays where each sub-array contains items for which
+ *    the String-returning function has returned the same value.
+ * @see R.groupBy
+ * @example
+ *      R.collectBy(R.prop('type'), [
+ *        {type: 'breakfast', item: '☕️'},
+ *        {type: 'lunch', item: '🌯'},
+ *        {type: 'dinner', item: '🍝'},
+ *        {type: 'breakfast', item: '🥐'},
+ *        {type: 'lunch', item: '🍕'}
+ *      ]);
+ *
+ *      // [ [ {type: 'breakfast', item: '☕️'},
+ *      //     {type: 'breakfast', item: '🥐'} ],
+ *      //   [ {type: 'lunch', item: '🌯'},
+ *      //     {type: 'lunch', item: '🍕'} ],
+ *      //   [ {type: 'dinner', item: '🍝'} ] ]
+ */
+var collectBy = _curry2(function collectBy(fn, list) {
+  return values(groupBy(fn, list));
```

Comment:
If you have some time, could you compare the performance to a iteration-based approach?  There may not be much difference, in which case, we can keep this simple code.  But if the difference is substantial (perhaps: a factor of more than 2), we should probably switch to more performant code.

Reply:
How do you usually do this? Do you use any benchmarking library?

Reply:
If a grouping function returns a non-string value must cause `collectBy` to return an empty list, then I'd probably have to implement with classic loop statements. Do you still want to me to measure the performance in that case and what would be an acceptable threshold?

Reply:
> Do you use any benchmarking library?

Usually a quick test in [jsperf][jp] or [JSBench.me][jb] is plenty.  This doesn't have to live long.

  [jb]:https://jsbench.me/
  [jp]: https://jsperf.com/

Reply:
I would simply use the default object behavior for keys, which should take no effort: 

```js
const foo = {}
foo[true] = 1
foo[{a: 1}] = 2
foo[/[a-z]/] = 3
foo[(n) => n * n] = 4
// etc.

foo //=>
// {
//   "/[a-z]/": 3,
//   "[object Object]": 2,
//   "function (n) {\n  return n * n;\n}": 4, // (or whatever)
//   "true": 1
// }
```

If you peruse our source code, many functions which could easily be written in terms of other ones are written natively, often with `while` loops.  The rationale is that the library does this nasty stuff so you don't have to.

Reply:
I ended up writing the function "natively". Would you still see some performance benchmark?

Reply:
No.  The only reason for that was that if we were skipping our usual style, I wanted to ensure that it didn't hurt performance.